### ISSUE TITLE
Explicit panic when submitting command buffer

### ISSUE THREAD

**Issue by dobrakmato** [vulkano-rs/vulkano#1560]
# Template

*   Version of vulkano: 0.21.0
*   OS: Windows
*   GPU (the selected PhysicalDevice): GTX 1060

# Issue

I am having a strange issue with triggering `panic!()` call on this line https://github.com/vulkano-rs/vulkano/blob/master/vulkano/src/command_buffer/synced/commands.rs#L2683. 

> thread 'main' panicked at 'explicit panic', vulkano\src\command_buffer\synced\commands.rs:2272:25

I do not understand the vulkano internals very well, but I believe this part of code is supposed to track usage of images in recorded buffer and verify that they are used correctly (with respect to memory).

I also cannot reproduce the bug consistently in every run of the application, which may suggest that this is some kind of race condition bug. 

From what I gathered using the debugger the `image` function is usually called with low numbers such as 0, 1 or 2. But in sometimes it happens to be called with unusually high number (eg. 5), which triggers the `panic` call.

I am able to trigger this by alt-tabing from and into the application's window, but this is also not consistent. I am able to trigger it more consistently by resizing the application's window, however it happens without having to resize the window (so I think it is not related to how I handle the resize). 

As I am currently unable to find the cause of the issue I cannot provide MVP, so this is more of a question what would be the best way to debug this?

My render pipeline consists of two render passes. First is deferred render to HDR buffer (geometry pass and lighting pass) followed by skybox pass and another pass which resolves the HDR buffer to LDR buffer with tone mapping. 

```rust
  let render_pass = Arc::new(
            vulkano::ordered_passes_renderpass!(
                device.clone(),
                attachments: {
                    gbuffer1: {
                        load: Clear,
                        store: Store,
                        format: Format::A2B10G10R10UnormPack32,
                        samples: 1,
                    },
                    gbuffer2: {
                        load: Clear,
                        store: Store,
                        format: Format::R8G8B8A8Unorm,
                        samples: 1,
                    },
                    gbuffer3: {
                        load: Clear,
                        store: Store,
                        format: Format::R8G8B8A8Unorm,
                        samples: 1,
                    },
                    depth: {
                        load: Clear,
                        store: DontCare,
                        format: Format::D16Unorm,
                        samples: 1,
                    },
                    hdr: {
                        load: Clear,
                        store: DontCare,
                        format: Format::B10G11R11UfloatPack32,
                        samples: 1,
                    },
                    ldr: {
                        load: DontCare,
                        store: Store,
                        format: Format::B10G11R11UfloatPack32,
                        samples: 1,
                    }
                },
                passes: [
                    {
                        color: [gbuffer1, gbuffer2, gbuffer3],
                        depth_stencil: {depth},
                        input: []
                    },
                    {
                        color: [hdr],
                        depth_stencil: {},
                        input: [gbuffer1, gbuffer2, gbuffer3, depth]
                    },
                    {
                        color: [hdr],
                        depth_stencil: {depth},
                        input: []
                    },
                    {
                         color: [ldr],
                         depth_stencil: {},
                         input: [hdr]
                    }
                ]
            )
            .expect("cannot create render pass"),
        );
```

And the second render pass takes the LDR buffer produced by first pass and applies FXAA on it and writes the results into window framebuffer. 

```rust
   let render_pass = Arc::new(
            vulkano::ordered_passes_renderpass!(
                device.clone(),
                attachments: {
                    final_color: {
                        load: DontCare,
                        store: Store,
                        format: swapchain_format,
                        samples: 1,
                    }
                },
                passes: [
                    {
                         color: [final_color],
                         depth_stencil: {},
                         input: []
                    }
                ]
            )
            .expect("cannot create render pass for fxaa"),
        );
```

I am attaching my render loop in case I missed something trivial I should do. The `render_frame` function is called from winit ` Event::RedrawEventsCleared` event inside the event-loop.

```rust
   pub fn render_frame(&mut self, game_state: &GameState) {
        self.frame_no += 1;

        // if framebuffers are out-of date, we need to recreate them.
        if self.framebuffers_out_of_date {
            self.recreate_framebuffers();
            self.framebuffers_out_of_date = false;
            return;
        }

        // clean-up all resources from the previous frame
        if let Some(t) = self.previous_frame_end.as_mut() {
            t.cleanup_finished();
        }

        // acquire next image from swapchain that will be used for rendering. if the
        // suboptimal flag is true we try to recreate the swapchain after this frame.
        //
        // if the acquire operation fails, we recreate swapchain right away and skip
        // rendering of this frame
        let (idx, suboptimal, acquire_future) =
            match swapchain::acquire_next_image(self.swapchain.clone(), None) {
                Ok(r) => r,
                Err(e) => {
                    warn!("Cannot acquire next image {:?}. Recreating swapchain...", e);
                    self.recreate_swapchain();
                    return;
                }
            };

        // build primary command buffer by distributing command buffer
        // recording into multiple threads as parallel job
        let mut frame = Frame {
            render_path: &mut self.render_path,
            game_state,
            framebuffer: self.framebuffers[idx].clone(),
            builder: Some(
                AutoCommandBufferBuilder::primary_one_time_submit(
                    self.device.clone(),
                    self.graphical_queue.family(),
                )
                .unwrap(),
            ),
        };

        // let frame create and records it's command buffer(s).
        let primary_cb = frame.build();

        // wait for image to be available and then present drawn the image
        // to screen.
        let future = self
            .previous_frame_end
            .take()
            .unwrap()
            .join(acquire_future)
            .then_execute(self.graphical_queue.clone(), primary_cb)
            .unwrap()
            .then_swapchain_present(self.graphical_queue.clone(), self.swapchain.clone(), idx)
            .then_signal_fence_and_flush();

        // depending on the completion state of the submitted command buffer either
        // return to continue to next frame, or report and error
        match future {
            Ok(f) => {
                self.previous_frame_end = Some(f.boxed());
            }
            Err(e) => {
                error!("Error occurred during rendering a frame {:?}", e);
                self.previous_frame_end = now(self.device.clone());
            }
        }

        // if we hit the `suboptimal` flag, recreate the swapchain now, after
        // the rendering work has been submitted
        if suboptimal {
            warn!("Swapchain is suboptimal! Recreating swapchain.");
            self.recreate_swapchain();
        }
    }
```

**Comment by Rua** [comment]
The way `SyncCommandBuffer` works, is that when building, commands that use resources report each resource in sequence, which is then given a number based on its place in the sequence. That `image` function is then called to retrieve the resource later, by giving it the number of the resource you want. In this case, it's being called on the command for binding descriptor sets, and `image` is given a number that's out of range.

Why that's happening I can't tell off-hand. But what you can try is upgrading to 0.22 or the latest master version, and see if the problem still occurs then. If it does still occur, could you make some minimal working code that can reproduce the problem? Then I can have a look.

**Comment by dobrakmato** [comment]
> The way SyncCommandBuffer works, is that when building, commands that use resources report each resource in sequence, which is then given a number based on its place in the sequence. That image function is then called to retrieve the resource later, by giving it the number of the resource you want. In this case, it's being called on the command for binding descriptor sets, and image is given a number that's out of range.

What is troubling me is that is happens only sometimes. My guess was that the buffers are still used for rendering in n-th frame when n+1 th frame is sent to GPU, but I am getting this error even when I `.wait()` the future from `then_signal_fence_and_flush`.


After upgrading to  0.22 I am getting `PersistentDescriptorSetError::EmptyExpected` error in a different part of code. I guess I will keep debugging to first fix that and maybe this issue disappears.

**Comment by Rua** [comment]
The panic is probably happening when you submit the command buffer, regardless of what's going on elsewhere. Submitting a command buffer will try to lock all the resources it uses in `SyncCommandBuffer::lock_submit`, which uses `buffer` and `image` to retrieve each resource. And if the index for an image is out of range it panics. So something in the logic of `SyncCommandBuffer` is producing invalid image indices...

**Comment by dobrakmato** [comment]
Thanks for the response! 

I found out that I also have a problem with `readonly` flag being `false` on uniform buffers in code generated from `shader!` macro. I will port my whole project to vulkano's master branch first.

I also believe that `panic!` call is supposed to represent unreachable code (under normal circumstances) as it is called from `eq` function during `HashMap::get` call. I will try to look into why it produced invalid index and that will hopefully solve my issue.

**Comment by funmaker** [comment]
I've encountered similar issue in my project and I think the root cause here is the same.

The issue here is `SyncCommandBufferBuilder` assuming that each resource will have the same index for every command in command buffer. Function `SyncCommandBufferBuilder::append_command` appends resources used by a new command that is being inserted. If the resource was already used, it simply appends new command id to the existing `BuilderKey` but the existing key might have different `resource_index`. This seems pretty bad, for large projects this might cause seemingly random panics and unreliable resource handling.

I've made smaller, consistent reproduction: funmaker/vulkano-issue-1560
The reason this bug is hard to reproduce is because `BuilderKey`s are rarely compared because they are stored in a hashmap. I've included a patch in that repo to compare new key to all the keys in the builder to make this issue manifest consistently.

As for fix, I think BuilderKey and CbKey should hold a vector of pairs of command id and resource indexes, instead of one common `resource_index` for all the commands in key.

**Comment by nxfs** [comment]
I am hitting the same error with a similar setup using two render passes, on vulkano 0.24.
Here's the backtrace:

```
0: std::panicking::begin_panic                                                                                                                                                                                                                                                         
             at /rustc/607d6b00d4e0e0475b8de9d0c870b7126fdcdf6b/library/std/src/panicking.rs:541:12                                                                                                                                                                                       
   1: <<vulkano::command_buffer::synced::commands::SyncCommandBufferBuilderBindDescriptorSets::submit::Cmd<I> as vulkano::command_buffer::synced::base::Command>::into_final_command::Fin as vulkano::command_buffer::synced::base::FinalCommand>::image                                  
             at <redacted>/.cargo/registry/src/github.com-1ecc6299db9ec823/vulkano-0.24.0/src/command_buffer/synced/commands.rs:2703:25                                                                                                                                 
   2: <vulkano::command_buffer::synced::base::CbKey as core::cmp::PartialEq>::eq::{{closure}}                                                                                                                                                                                             
             at <redacted>/.cargo/registry/src/github.com-1ecc6299db9ec823/vulkano-0.24.0/src/command_buffer/synced/base.rs:1486:69
   3: <core::slice::iter::Iter<T> as core::iter::traits::iterator::Iterator>::any                                                            
             at /rustc/607d6b00d4e0e0475b8de9d0c870b7126fdcdf6b/library/core/src/slice/iter/macros.rs:234:24                                 
   4: <vulkano::command_buffer::synced::base::CbKey as core::cmp::PartialEq>::eq                                                                                                                                                                                                          
             at <redacted>/.cargo/registry/src/github.com-1ecc6299db9ec823/vulkano-0.24.0/src/command_buffer/synced/base.rs:1484:37
   5: hashbrown::map::equivalent_key::{{closure}}                                                                                            
             at /cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.11.0/src/map.rs:233:14
   6: hashbrown::raw::RawTable<T,A>::find                                                                                                    
             at /cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.11.0/src/raw/mod.rs:923:27
   7: hashbrown::raw::RawTable<T,A>::get_mut                                                                                                                                                                                                                                              
             at /cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.11.0/src/raw/mod.rs:945:15
   8: hashbrown::map::HashMap<K,V,S,A>::insert                                                                                                                                                                                                                                            
             at /cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.11.0/src/map.rs:1263:34                                          
   9: <hashbrown::map::HashMap<K,V,S,A> as core::iter::traits::collect::Extend<(K,V)>>::extend::{{closure}}                                  
             at /cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.11.0/src/map.rs:3405:13                                          
  10: core::iter::traits::iterator::Iterator::for_each::call::{{closure}}                                                                    
             at /rustc/607d6b00d4e0e0475b8de9d0c870b7126fdcdf6b/library/core/src/iter/traits/iterator.rs:733:29
  11: core::iter::adapters::map::map_fold::{{closure}}                                                                                       
             at /rustc/607d6b00d4e0e0475b8de9d0c870b7126fdcdf6b/library/core/src/iter/adapters/map.rs:82:21
  12: core::iter::traits::iterator::Iterator::fold                                                                                           
             at /rustc/607d6b00d4e0e0475b8de9d0c870b7126fdcdf6b/library/core/src/iter/traits/iterator.rs:2153:21                             
  13: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold                                                  
             at /rustc/607d6b00d4e0e0475b8de9d0c870b7126fdcdf6b/library/core/src/iter/adapters/map.rs:122:9
  14: core::iter::traits::iterator::Iterator::for_each                                                                                       
             at /rustc/607d6b00d4e0e0475b8de9d0c870b7126fdcdf6b/library/core/src/iter/traits/iterator.rs:736:9                               
  15: <hashbrown::map::HashMap<K,V,S,A> as core::iter::traits::collect::Extend<(K,V)>>::extend                                               
             at /cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.11.0/src/map.rs:3404:9                                           
  16: <std::collections::hash::map::HashMap<K,V,S> as core::iter::traits::collect::Extend<(K,V)>>::extend     
             at /rustc/607d6b00d4e0e0475b8de9d0c870b7126fdcdf6b/library/std/src/collections/hash/map.rs:2780:9                               
  17: <std::collections::hash::map::HashMap<K,V,S> as core::iter::traits::collect::FromIterator<(K,V)>>::from_iter                           
             at /rustc/607d6b00d4e0e0475b8de9d0c870b7126fdcdf6b/library/std/src/collections/hash/map.rs:2765:9
  18: core::iter::traits::iterator::Iterator::collect                                                                                        
             at /rustc/607d6b00d4e0e0475b8de9d0c870b7126fdcdf6b/library/core/src/iter/traits/iterator.rs:1748:9                              
  19: vulkano::command_buffer::synced::base::SyncCommandBufferBuilder::build                                                                                                                                                                                                              
             at <redacted>/.cargo/registry/src/github.com-1ecc6299db9ec823/vulkano-0.24.0/src/command_buffer/synced/base.rs:940:13                                                                                                                                      
  20: vulkano::command_buffer::auto::AutoCommandBufferBuilder<vulkano::command_buffer::auto::PrimaryAutoCommandBuffer<<P as vulkano::command_buffer::pool::CommandPoolBuilderAlloc>::Alloc>,P>::build                                                                                     
             at <redacted>/.cargo/registry/src/github.com-1ecc6299db9ec823/vulkano-0.24.0/src/command_buffer/auto.rs:409:20 
```

**Comment by nxfs** [comment]
> As for fix, I think BuilderKey and CbKey should hold a vector of pairs of command id and resource indexes, instead of just one resource_index shared across all the commands in key. But I'm not 100% sure that's all.

I've analyzed the code and came to the same conclusion. However the codebase is fairly complex so it's very possible that I am missing something.

I've created a patch that implements the solution above. It seems to be working. I've tested against unit tests, against the minimal example and against my project which exhibited the failure, all seem to work. I'll send the pull request shortly. Please provide feedback on the approach and on what other testing could be done.

**Comment by Rua** [comment]
#1639 may render your code obsolete, because it reworks the code significantly and eliminates `BuilderKey` and `CbKey` as types entirely. It would be worth investigating how your solution would be implemented on top of the new code.

**Comment by Eliah-Lakhin** [comment]
@dobrakmato We have recently merged #1639 that probably fixed your issue. Can you confirm it?

**Comment by dobrakmato** [comment]
@Eliah-Lakhin Yup seems to be fixed for me. Many thanks!
